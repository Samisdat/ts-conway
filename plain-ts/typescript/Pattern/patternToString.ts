import {Pattern} from '@Conway/Pattern/Pattern';
import {matrixToString} from '@Conway/Geometry/matrixToString';

export const patternToString = (pattern: Pattern): string => {

    let patternAsString = '!Name: ' + pattern.getName() + '\n';

    patternAsString = patternAsString + matrixToString(pattern.getMatrix(), false)

    return patternAsString;

}
