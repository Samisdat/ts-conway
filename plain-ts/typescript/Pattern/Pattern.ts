import {Position} from '@Conway/Geometry/Position';
import {readPatternFromPlainFile} from '@Conway/Pattern/readPatternFromPlainFile';
import {Matrix} from '@Conway/Geometry/Matrix';
import {patternToString} from '@Conway/Pattern/patternToString';

export class Pattern {

    private name: string;

    private matrix: Matrix;

    constructor(name: string, matrix: Matrix) {

        this.name = name;

        this.matrix = matrix;

    }

    public static fromString(patternString: string): Pattern {
        return readPatternFromPlainFile(patternString);
    }

    public getName(): string {
        return this.name;
    }

    public getMatrix(): Matrix {

        return this.matrix;
    }

    public getWidth(): number {
        return this.matrix.getWidth();
    }

    public getHeight(): number {
        return this.matrix.getHeight();
    }

    public mirrorVertical(): void {

        const mirror = new Matrix();

        for (const position of this.matrix.getAll()) {

            mirror.add(
                new Position(
                    position.x,
                    -1 * position.y
                )
            );

        }

        this.matrix = mirror;

    }

    public mirrorHorizontal(): void {

        const mirror = new Matrix();

        for (const position of this.matrix.getAll()) {

            mirror.add(
                new Position(
                    -1 * position.x,
                    position.y
                )
            );

        }

        this.matrix = mirror;

    }

    public rotate(): void {

        const matrix = this.toArray();

        const rotated: number[][] = [];

        for (let x = 0; x < matrix[0].length; x += 1) {

            const row: number[] = [];

            for (let y = matrix.length - 1; y >= 0 ; y -= 1) {

                row.push(matrix[y][x]);

            }

            rotated.push(row);

        }
        const rotatedMatrix = new Matrix();

        for (let y = 0; y < rotated.length; y += 1) {

            for (let x = 0; x < rotated[y].length; x += 1) {

                if (0 === rotated[y][x]) {
                    continue;
                }

                rotatedMatrix.add(new Position(x, y));

            }

        }

        this.matrix = rotatedMatrix;

    }

    public toArray(): number[][] {

        const moved = new Matrix();

        const moveBy = this.getMatrix().getBound().bottomLeft().inverse();

        for (const position of this.matrix.getAll()) {

            moved.add(
                position.move(moveBy)
            );

        }

        const matrixAsArray: number[][] = [];

        for (let y = 0; y < this.getHeight() + 1; y += 1) {

            const row: number[] = [];

            for (let x = 0; x < this.getWidth() + 1; x += 1) {

                const cell: number = (true === moved.has(new Position(x, y))) ? 1 : 0;

                row.push(cell);

            }

            matrixAsArray.push(row);
        }

        return matrixAsArray;

    }

    public toString(): string {

        return patternToString(this);

    }

}
