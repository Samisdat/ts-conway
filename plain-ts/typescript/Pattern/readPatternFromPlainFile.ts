import {Position} from '@Conway/Geometry/Position';
import {Pattern} from '@Conway/Pattern/Pattern';
import {Matrix} from '@Conway/Geometry/Matrix';

interface ExtractedPattern {
    name: string;
    pattern: string[];
}

const extractePattern = (patternPlainText: string): ExtractedPattern => {

    const extractedPattern: ExtractedPattern = {
        name: 'unnamed',
        pattern: []
    };

    const rows: string[] = patternPlainText.split('\n').reverse();

    const isComment = new RegExp('^!(.*?)');
    const isMeta = new RegExp('^!(.*?)\:(.*?)$');

    for (const row of rows) {

        if (true === isComment.test(row)) {

            if (true === isMeta.test(row)) {

                const meta = isMeta.exec(row) as RegExpMatchArray;

                if ('Name' === meta[1]) {
                    extractedPattern.name = meta[2].trim();
                }

            }

        } else {

            extractedPattern.pattern.push(row);

        }

    }

    return extractedPattern;

};

export const readPatternFromPlainFile = (patternPlainText: string): Pattern => {

    const extractedPattern = extractePattern(patternPlainText);

    const matrix = new Matrix();

    let y = 0;
    let x = 0;

    for (const row of extractedPattern.pattern) {

        const cols: string[] = row.trim().split('');

        x = 0;

        for (const col of cols) {

            if ('O' === col) {

                matrix.add(new Position(x, y));

            }

            x += 1;
        }

        y += 1;

    }

    const pattern = new Pattern(extractedPattern.name, matrix);

    return pattern;

};

