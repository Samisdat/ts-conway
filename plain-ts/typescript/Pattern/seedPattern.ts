
import {Pattern} from '@Conway/Pattern/Pattern';
import {Position} from '@Conway/Geometry/Position';
import {Matrix} from '@Conway/Geometry/Matrix';

export const seedPattern = (matrix: Matrix, pattern: Pattern, moveBy = new Position(0, 0)): void => {

    const positions = pattern.getMatrix().getAll();

    const offSet = new Position(
        -1 * Math.floor(pattern.getMatrix().getWidth() / 2),
        -1 * Math.floor(pattern.getMatrix().getHeight() / 2)
    ).move(moveBy);

    for (const position of positions) {

        matrix.add(position.move(moveBy));

    }

}
