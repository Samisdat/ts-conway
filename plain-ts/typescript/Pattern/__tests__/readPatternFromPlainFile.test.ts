import serializer from '../../../jest-serialize-conway';
import {readPatternFromPlainFile} from '@Conway/Pattern/readPatternFromPlainFile';
import {glider} from '@Conway/Pattern/Store/glider.cells';
import {Pattern} from '@Conway/Pattern/Pattern';


describe('PatternFromPlainText', () => {

    beforeAll(() => {
        expect.addSnapshotSerializer(serializer);
    });

    test('can be created with a complete *.cell file', () => {

        const pattern = readPatternFromPlainFile(glider);

        expect(pattern).toBeInstanceOf(Pattern);

        expect(pattern.getMatrix()).toMatchSnapshot();

    });

    test('can be created with a .cell without meta ', () => {

        const patternStr = 'O..\n' +
            'OOO';

        const scholar = readPatternFromPlainFile(patternStr);

        expect(scholar).toBeInstanceOf(Pattern);

        expect(scholar.getMatrix()).toMatchSnapshot();

    });

});
