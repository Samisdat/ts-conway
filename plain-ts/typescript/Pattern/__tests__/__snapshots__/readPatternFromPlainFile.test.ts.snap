// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PatternFromPlainText can be created with a .cell without meta  1`] = `
XOO
XXX
`;

exports[`PatternFromPlainText can be created with a complete *.cell file 1`] = `
OXO
OOX
XXX
`;
